### PR TITLE
convert to use new adapter API

### LINKED ISSUE #29: update to use new adapter API

### PR DESCRIPTION
**Pull request by rlgomes**
fixes #29

this is a huge overhaul of things to use the new adapter API which also
gives us the ability to do "live" data reading.

I also piggy backed on this change the move to use eslint and gulp to
run things

### CODE REVIEW

**Comment by rlgomes** [comment]
@demmer @bkutil @davidvgalbraith I can split this up if we REALLY need to keep the eslint + gulp stuff in a separate PR but it was just easier to get things done in one shot. Other than that its just moving bits around to use the new adapter API which gives us a few niceties for free. 

@demmer @davidvgalbraith what is the right way to change the default lag ? I'd like to make it 5s here or otherwise maybe we should consider setting lag to 5s by default as I don't see a lag of 1s working with most "big data adapters" ?

**Comment by dmehra** [comment]
Sounds like you're asking for https://github.com/juttle/juttle/issues/308 - should we prioritize it?

File: lib/graphite.js [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        var auth = new buffer.Buffer(config.webapp.username + ':' +
+                                     config.webapp.password).toString('base64');
+        this.headers = {
+            'Authorization': 'Basic ' + auth.toString('base64')
+        };
+        this.config = config;
+    },
+
+    get: function(name, from, to) {
+        // comment missing
```
Comment (davidvgalbraith) [inline lib/graphite.js]: Comment missing?

File: lib/graphite.js [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        var fromSec = Math.round(new JuttleMoment(from).milliseconds()/1000);
```
Comment (davidvgalbraith) [inline lib/graphite.js]: `new JuttleMoment(from).seconds()` is probably what you want here.

File: lib/read.js [hunk 3]
```diff
@@ -1,130 +1,97 @@
...
+var AdapterRead = require('juttle/lib/runtime/adapter-read');
+var graphite = require('./graphite');
+var JuttleMoment = require('juttle/lib/runtime/types').JuttleMoment;
 
-            var from = Math.round(new JuttleMoment(this.from).milliseconds()/1000);
-            var to;
+class ReadGraphite extends AdapterRead {
+    constructor(options, params) {
+        super(options, params);
+        this.procName = 'read graphite';
```
Comment (davidvgalbraith) [inline lib/read.js]: Don't think adapters have `procName`s anymore.

File: lib/read.js [hunk 4]
```diff
@@ -1,130 +1,97 @@
...
+        var name = null;
 
-            var auth = new buffer.Buffer(config.webapp.username + ':' + config.webapp.password).toString('base64');
-            this.headers = {
-                'Authorization': 'Basic ' + auth.toString('base64')
-            };
-
-            var name = null;
-            if (params.filter_ast) {
+        if (params.filter_ast) {
```
Comment (davidvgalbraith) [inline lib/read.js]: Seems weird to pass the name as a filter instead of a `-name` option.

File: lib/read.js [hunk 5]
```diff
@@ -1,130 +1,97 @@
...
+    }
+
+    defaultTimeRange() {
+        return {
+            from: new JuttleMoment(0),
+            to: new JuttleMoment(Infinity)
+        };
+    }
+
+    read(from, to, limit, state) {
```
Comment (davidvgalbraith) [inline lib/read.js]: You totally disregard `limit`, should we have some enforcement for it?

File: lib/read.js [hunk 6]
```diff
@@ -1,130 +1,97 @@
...
 
-module.exports = Read;
+            this.parseTime(points, this.timeField);
+
+            return {
+                points: points,
+                readEnd: to
+            };
+        })
+        .catch((err) => {
```
Comment (davidvgalbraith) [inline lib/read.js]: The convention is you just throw fatal errors / reject the promise returned by `read`, returning data as if everything's fine is a risky proposition.

File: lib/write.js [hunk 7]
```diff
@@ -1,70 +1,61 @@
...
-        initialize: function(options, params) {
-            // XXX: missing checks for all required config values
-            this.name = 'writex-graphite';
-            var self = this;
+        this.procName = 'writex-graphite';
 
-            this.socket = net.Socket();
-            this.socket.connect(config.carbon.port, config.carbon.host);
+        // XXX: missing checks for all required config values
+        this.name = 'writex-graphite';
```
Comment (davidvgalbraith) [inline lib/write.js]: `writex` is sooo 2015.

File: lib/write.js [hunk 8]
```diff
@@ -1,70 +1,61 @@
...
-        },
+        this.socket.on('end', function() {
+            // only eof when we've closed the socket
+            self.on_eof(self);
+        });
 
-        process: function(points) {
-            var self = this;
-            var payload = [];
+        this.isDone = new Promise(function(resolve, reject) {
```
Comment (davidvgalbraith) [inline lib/write.js]: Think we got rid of the silly `isDone` business.

File: lib/write.js [hunk 9]
```diff
@@ -1,70 +1,61 @@
...
+        self.socket.write(payload.join(''), 'utf-8', function(err) {
+            if (err) {
+                self.trigger('error', err);
+                self.done();
+            }
+        });
+    }
+
+    eof() {
+        this.socket.end();
```
Comment (davidvgalbraith) [inline lib/write.js]: Are we synchronously done after `this.socket.end()` or is there any cleanup we should be waiting for?

File: lib/read.js (hunk 4 above)
Comment (rlgomes) [inline lib/read.js]: This was done before and then @demmer suggested we do it based on the filter expression which I have to say is "easier" to remember than whats the name of that flag again ? Also allows us to express things such as `name=x` vs `name~x.*` which graphite supports and later we could even support OR filters where we'd issue 2 queries on your behalf if you did something like `name=A OR name=B` ...

File: lib/read.js (hunk 5 above)
Comment (rlgomes) [inline lib/read.js]: I haven't seen limit enforcement in any of the built-ins so I wasn't sure what to do wit it. As you can see: 

https://github.com/juttle/juttle/blob/master/lib/adapters/file/read.js#L60
https://github.com/juttle/juttle/blob/master/lib/adapters/stdio/read.js#L46
https://github.com/juttle/juttle/blob/master/lib/adapters/http/read.js#L146

So what does limit mean ? the maximum amount allowed to read or the maximum we should return in each return from read ?

File: lib/read.js (hunk 6 above)
Comment (rlgomes) [inline lib/read.js]: This was here and I thought it looked odd so I will no longer return any data which I agree with you makes more sense and just throw the error.

File: lib/write.js (hunk 9 above)
Comment (rlgomes) [inline lib/write.js]: will update this to actually return a promise that only resolves once the socket has really ended...

**Comment by rlgomes** [comment]
bump

**Comment by davidvgalbraith** [comment]
+1

**Pull request merged**